File: HermesMobile/LiveActivities/HermesDeepLink.swift
import Foundation

enum HermesDeepLink {
    static var scheme: String {
        Bundle.main.object(forInfoDictionaryKey: "HermesURLScheme") as? String
            ?? "hermes-agent"
    }

    static let sessionHost = "session"

    /// Host for the one bot route (#554): `hermes-agent://bot?server=…&connection=…&profile=…`.
    /// The builder, parser and typed `BotDestination` live in `Features/Bots/BotDeepLink.swift`,
    /// which is main-app only; this file is shared with the Live Activity widget.
    static let botHost = "bot"

    /// Host for the parameter-less "open the New Chat composer" deep link used by the
    /// New Chat App Intent (issue #337). Mirrors the share extension's host-based routing
    /// so the intent can reuse `ContentView.handleOpenURL` rather than inventing a new path.
    static let newChatHost = "new-chat"

    /// `hermes-agent://new-chat` (scheme follows the active build, e.g. `-branch`).
    static var newChatURL: URL? {
        var components = URLComponents()
        components.scheme = scheme
        components.host = newChatHost
        return components.url
    }

    static func isNewChatURL(_ url: URL) -> Bool {
        url.scheme?.lowercased() == scheme
            && url.host?.lowercased() == newChatHost
    }

    /// Host for "open the New Chat composer *and* auto-start voice dictation", used by the
    /// "New Chat with Voice" App Intent (issue #338). A distinct host from `newChatHost`
    /// so the two intents never alias each other — `isNewChatURL` and `isNewChatVoiceURL`
    /// are mutually exclusive.
    static let newChatVoiceHost = "new-chat-voice"

    /// `hermes-agent://new-chat-voice` (scheme follows the active build, e.g. `-branch`).
    static var newChatVoiceURL: URL? {
        var components = URLComponents()
        components.scheme = scheme
        components.host = newChatVoiceHost
        return components.url
    }

    static func isNewChatVoiceURL(_ url: URL) -> Bool {
        url.scheme?.lowercased() == scheme
            && url.host?.lowercased() == newChatVoiceHost
    }

    /// Host for "open the New Chat composer pinned to a specific profile", used by the
    /// "New Chat in <Profile>" App Intent (issue #339). A distinct host from the other
    /// new-chat hosts so the three intents never alias; the profile name rides as a query
    /// item (like `sessionURL`'s `id`) rather than in the host, so it can carry spaces and
    /// non-ASCII safely via percent-encoding.
    static let newChatInProfileHost = "new-chat-profile"

    /// Query-item name carrying the profile's server name.
    static let profileQueryItem = "profile"

    /// `hermes-agent://new-chat-profile?profile=<name>` (scheme follows the active build).
    /// Returns nil for a blank profile name so callers can pass it straight through.
    static func newChatInProfileURL(profileName: String) -> URL? {
        let trimmed = profileName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        var components = URLComponents()
        components.scheme = scheme
        components.host = newChatInProfileHost
        components.queryItems = [URLQueryItem(name: profileQueryItem, value: trimmed)]
        return components.url
    }

    static func isNewChatInProfileURL(_ url: URL) -> Bool {
        url.scheme?.lowercased() == scheme
            && url.host?.lowercased() == newChatInProfileHost
    }

    /// Extracts the profile name from a "New Chat in <Profile>" URL, or nil when the URL is a
    /// different kind or carries no (non-blank) profile.
    static func profileName(fromNewChatInProfile url: URL) -> String? {
        guard isNewChatInProfileURL(url) else { return nil }

        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        guard let raw = components?.queryItems?.first(where: { $0.name == profileQueryItem })?.value
        else {
            return nil
        }

        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    static func sessionURL(sessionID: String) -> URL? {
        guard !sessionID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }

        var components = URLComponents()
        components.scheme = scheme
        components.host = sessionHost
        components.queryItems = [
            URLQueryItem(name: "id", value: sessionID)
        ]
        return components.url
    }

    static func sessionID(from url: URL) -> String? {
        guard url.scheme?.lowercased() == scheme,
              url.host?.lowercased() == sessionHost
        else {
            return nil
        }

        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        if let id = components?.queryItems?.first(where: { item in
            item.name == "id" || item.name == "session_id"
        })?.value {
            return normalizedSessionID(id)
        }

        let pathID = url.pathComponents
            .filter { $0 != "/" }
            .first
        return normalizedSessionID(pathID)
    }

    private static func normalizedSessionID(_ rawValue: String?) -> String? {
        let trimmed = rawValue?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? nil : trimmed
    }
}

File: app/src/main/java/com/hermex/app/ui/navigation/HermesDeepLink.kt
package com.hermex.app.ui.navigation

import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Locale

/** Host-based deep-link parser mirroring the iOS HermesDeepLink contract. */
object HermesDeepLink {
    const val SCHEME = "hermes-agent"
    private const val SESSION_HOST = "session"
    private const val NEW_CHAT_HOST = "new-chat"
    private const val NEW_CHAT_VOICE_HOST = "new-chat-voice"
    private const val NEW_CHAT_PROFILE_HOST = "new-chat-profile"
    private const val PROFILE_QUERY_ITEM = "profile"

    fun newChatUrl(): String = "$SCHEME://$NEW_CHAT_HOST"
    fun newChatVoiceUrl(): String = "$SCHEME://$NEW_CHAT_VOICE_HOST"
    fun newChatInProfileUrl(profileName: String): String? {
        val trimmed = profileName.trim()
        if (trimmed.isEmpty()) return null
        return "$SCHEME://$NEW_CHAT_PROFILE_HOST?$PROFILE_QUERY_ITEM=${encode(trimmed)}"
    }

    fun sessionUrl(sessionId: String): String? {
        val trimmed = sessionId.trim()
        if (trimmed.isEmpty()) return null
        return "$SCHEME://$SESSION_HOST?id=${encode(trimmed)}"
    }

    fun isNewChatUrl(url: String): Boolean = parse(url)?.let { uri ->
        uri.scheme?.lowercase(Locale.ROOT) == SCHEME && uri.host?.lowercase(Locale.ROOT) == NEW_CHAT_HOST
    } ?: false

    fun isNewChatVoiceUrl(url: String): Boolean = parse(url)?.let { uri ->
        uri.scheme?.lowercase(Locale.ROOT) == SCHEME && uri.host?.lowercase(Locale.ROOT) == NEW_CHAT_VOICE_HOST
    } ?: false

    fun isNewChatInProfileUrl(url: String): Boolean = parse(url)?.let { uri ->
        uri.scheme?.lowercase(Locale.ROOT) == SCHEME && uri.host?.lowercase(Locale.ROOT) == NEW_CHAT_PROFILE_HOST
    } ?: false

    fun profileNameFromNewChatInProfile(url: String): String? {
        if (!isNewChatInProfileUrl(url)) return null
        val raw = queryParameter(url, PROFILE_QUERY_ITEM)?.trim().orEmpty()
        return raw.takeIf { it.isNotEmpty() }
    }

    fun sessionId(url: String): String? {
        val uri = parse(url) ?: return null
        if (uri.scheme?.lowercase(Locale.ROOT) != SCHEME || uri.host?.lowercase(Locale.ROOT) != SESSION_HOST) return null
        val queryId = queryParameter(url, "id") ?: queryParameter(url, "session_id")
        if (!queryId.isNullOrBlank()) return queryId.trim()
        return uri.path
            ?.split('/')
            ?.firstOrNull { it.isNotBlank() }
            ?.trim()
            ?.takeIf { it.isNotEmpty() }
    }

    private fun parse(url: String): URI? = runCatching { URI(url) }.getOrNull()

    private fun queryParameter(url: String, name: String): String? {
        val rawQuery = parse(url)?.rawQuery ?: return null
        return rawQuery.split('&')
            .mapNotNull { pair ->
                val idx = pair.indexOf('=')
                if (idx < 0) null else pair.substring(0, idx) to pair.substring(idx + 1)
            }
            .firstOrNull { (key, _) -> decode(key) == name }
            ?.second
            ?.let(::decode)
    }

    private fun encode(value: String): String = java.net.URLEncoder.encode(value, StandardCharsets.UTF_8.name())
    private fun decode(value: String): String = URLDecoder.decode(value, StandardCharsets.UTF_8.name())
}
